EIPS/eip-7504.md

```diff
@@ -0,0 +1,421 @@
+
+| metadata.name           | The name of an Extension.                                             |
+| ----------------------- | --------------------------------------------------------------------- |
+| metadata.metadataURI    | A URI where we store some information about the Extension.            |
+| metadata.implementation | The implementation contract address of the Extension.                 |
+| metadata.functions      | All functions implemented in the Extension’s implementation contract. |
+
+The `Extension` struct is designed to let the smart contract and its clients help each other. Concretely, the design aims to create room for coordination between smart contracts and clients.
+
+- Naming extensions allows both parties to create a coherent, conceptual grouping of what function(-ality) lives together in which implementation smart contract.
+- An extension’s metadata URI allows for specific coordination between smart contracts and clients about how to understand, interact with or render UI for a given extension of a router.
+- Storing both function selectors **_and_** function signatures for each function implemented by an extension allows for a precise, joint ABI to be built for a dynamic contract.
+
+With this grouping of functions by Extensions, the `RouterState.getAllExtensions` function acts as a fixed function that returns a dynamic contract’s ABI.
+
+The list of extensions returned by `getAllExtensions` corroborates with the return values of `getImplementationForFunction`. Thus, these two functions taken together creates a reliable source to fetch and build a dynamic contract’s ABI.
+
+### Function selector clashes
+
+What happens when two separate implementation contracts used by the same dynamic contract implement the same function? (same function signature, even if the function bodies differ)
+
+This proposal requires that the fallback function must always perform a delegateCall to the return value of `getImplementationForFunction(msg.sig)`.
+
+This requirement must be met even in the situation described above. This proposal does not require a specific way of handling the situation.
+
+That said, it is _not_ necessary for the `Extension.functions` list, for a given implementation contract, to be an exhaustive list of all callable functions of the implementation contract.
+
+And so, in case two implementation contracts implement the same function, the `Extension.functions` list for one of the implementation contracts can simply omit the clashing function.
+
+We expand further on function selector clashes in the security considerations section, under “Accidental Upgrades”.
+
+### Adapters
+
+The focus of this proposal is comprehensibility and client friendliness for dynamic contracts.
+
+Contracts written in standards such as ERC-2535 or [ERC-6900](./eip-6900.md) which also focus on dynamic contracts, or ‘one-to-any’ proxy contracts can support this standard.
```

Comment:
Friendly warning: by mentioning ERC-6900 you create an implicit dependency from this EIP to that one. In practice, this means ERC-7504 will not be able to advance past ERC-6900's state (currently `Draft`.)

It's perfectly fine to keep the reference, just wanted to make sure you were aware of that rule.

Reply:
I have removed any mentions of ERC-6900. The earlier draft had meant to use ERC-6900 as an example, and did not intend to create any implicit dependency. Thanks for pointing this out.